runtime-modules/storage/src/lib.rs

```diff
@@ -2835,44 +2818,98 @@ impl<T: Trait> Module<T> {
         dynamic_bag_id: &DynamicBagId<T>,
         deletion_prize: &Option<DynamicBagDeletionPrize<T>>,
-    ) -> DispatchResult {
+        upload_params: &Option<UploadParameters<T>>,
+    ) -> Result<Option<BagUpdate<BalanceOf<T>>>, DispatchError> {
         let bag_id: BagId<T> = dynamic_bag_id.clone().into();
-
         ensure!(
             !<Bags<T>>::contains_key(bag_id),
             Error::<T>::DynamicBagExists
         );
 
-        if let Some(deletion_prize) = deletion_prize {
+        let bag_change = upload_params.as_ref().map_or(Ok(None), |params| {
+            // check params are valid
             ensure!(
-                Balances::<T>::usable_balance(&deletion_prize.account_id) >= deletion_prize.prize,
-                Error::<T>::InsufficientBalance
+                params.expected_data_size_fee == Self::data_object_per_mega_byte_fee(),
+                Error::<T>::DataSizeFeeChanged
             );
-        }
 
+            Self::check_global_uploading_block()?;
+
+            Self::ensure_objects_creation_list_validity(&params.object_creation_list)?;
+
+            // compute bag change due to objects creation list
+            Self::construct_bag_change(&params.object_creation_list).map(|res| Some(res))
+        })?;
+
+        let total_upload_fee = deletion_prize
+            .as_ref()
+            .map_or(Zero::zero(), |del_prize| del_prize.prize)
+            .saturating_add(bag_change.as_ref().map_or(Zero::zero(), |bag_change| {
+                Self::compute_upload_fees(bag_change)
+            }));
+
+        Self::ensure_sufficient_balance_for_upload(
+            deletion_prize
+                .as_ref()
+                .map(|deletion_prize| deletion_prize.account_id.clone()),
+            total_upload_fee,
+        )?;
+
+        Ok(bag_change)
+    }
+
+    fn ensure_sufficient_balance_for_upload(
+        deletion_prize_source_account_id: Option<T::AccountId>,
+        required_balance: BalanceOf<T>,
+    ) -> DispatchResult {
+        let usable_balance = deletion_prize_source_account_id.map_or(Zero::zero(), |account_id| {
+            Balances::<T>::usable_balance(account_id)
+        });
+
+        println!(
+            "TOTAL UPLOAD FEE: \t {:?}\nBALANCE: \t {:?}",
+            required_balance, usable_balance
+        );
+
+        ensure!(
+            usable_balance >= required_balance,
+            Error::<T>::InsufficientBalance
+        );
         Ok(())
     }
 
     // Validates dynamic bag deletion params and conditions. Returns bag's deletion prize.
     fn validate_delete_dynamic_bag_params(
         dynamic_bag_id: &DynamicBagId<T>,
+        with_objects: bool,
     ) -> Result<Option<BalanceOf<T>>, DispatchError> {
         Self::ensure_dynamic_bag_exists(dynamic_bag_id)?;
 
         let dynamic_bag = Self::dynamic_bag(dynamic_bag_id);
 
+        // deletion prize = bag.deletion_prize + total_objects fees if any
+        let deletion_prize = match with_objects {
```

Comment:
we should never match bool. Please, use if else in other cases. 

Here you can place ensure in 
```
if !with_objects {
   ensure!()
} 

let deletion_prize = .....
```